### ISSUE TITLE
Unable to run aurelia-skeleton-bridge

### ISSUE THREAD

**Issue by karthickthangasamy** [aurelia-ui-toolkits/skeleton-bridge#26]
I cloned the [aurelis-skeleton-bridge repo](https://github.com/aurelia-ui-toolkits/skeleton-bridge/) to my machine and executed the following commands.
-   npm install
-   jspm install
-   cd sample
-       npm install
-   jspm install

All the above steps went success. When I run **_gulp watch_**, I am facing the below issues.

**_Issue 1_**

`Uncaught (in promise) Error: XHR error (404 Not Found) loading http://localhost:3002/src/root/package.json(…)`

**_Issue 2_**

`system.src.js:4597 GET http://localhost:3002/src/skeleton-bridge.js 404 (Not Found)
Uncaught (in promise) Error: XHR error (404 Not Found) loading http://localhost:3002/src/skeleton-bridge.js(…)
`

**Comment by Thanood** [comment]
Strange, as [this line](https://github.com/aurelia-ui-toolkits/skeleton-bridge/blob/master/build/tasks/serve.js#L16) should setup your server-side routes to have a `/src/root` entry mapped to the project root.

This part does not seem to work in your case.
Just for trying it out: did you change the port yourself? Normally it's port 3000. If not, are you running another browserSync session on port 3000 which could conflict with the skeleton bridge app?

I don't know if it does, it's just a possibility. :smile: 
I'll try if the current version works for me. This might take a while.

**Comment by Thanood** [comment]
I'm getting the second error, too..

```
Failed to load resource: the server responded with a status of 404 (Not Found)
Uncaught (in promise) Error: XHR error (404 Not Found) loading http://localhost:3000/src/skeleton-bridge.js(…)
```

Correction: I'm getting an error (problably the first) here:

``` javascript
System.import("root/package.json!")
```

which leads to error 2.

I must be overlooking something totally obvious since this works in any other bridge I've tried yet. :smile:

**Comment by karthickthangasamy** [comment]
@Thanood Thanks. As per the discussion from this [thread](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/72).  We need to have the below setup in **index.html** file.

```
    System.import("root/package.json!").then(function(m) {
    var cfg = {
       packages: {}
    };

    // Create a new systemjs package with the package name
    cfg.packages[m.name] = {
        "main": "index",
          "dependencies": m.jspm.dependencies
        };

    System.config(cfg);

  });
```

**Comment by adriatic** [comment]
I am now repeating the steps @Karthick-NR just did. Hang on for a little while please:

### Confirmed - I am getting the same result as @karthickthangasamy

**Not sure why are you trying to use this repository** - as this is just a template to be used to create a new bridge project, so it might not be ready to run **as is**.

For completeness sake, I will not look into this issue and report my finding back here.

**Comment by Thanood** [comment]
@adriatic Last time I created a bridge it worked as is. I now compared the files I thought were essential  for this to work (build/serve configuration, index.html) and they are identical. i must be missing something.

As for SyncFusion, they have a working bridge. I was just able to clone and run it.

**Comment by adriatic** [comment]
### Problem solved - read on:

**[This line](https://github.com/aurelia-ui-toolkits/skeleton-bridge/blob/master/sample/src/main.js#L7)** should NOT be there - the correct version of `main.js` should be

```
import 'bootstrap';

export function configure(aurelia) {
  aurelia.use
    .standardConfiguration()
    .developmentLogging();

  aurelia.use.globalResources('shared/collapse-panel');
  aurelia.use.globalResources('shared/markdown');
  aurelia.use.globalResources('shared/logger');
  aurelia.use.globalResources('shared/au-code');

  aurelia.start()
    .then(au => au.setRoot('app'));
}

```

### Comment

Each Aurelia-UI-Tollkits **[catalog application](https://github.com/aurelia-ui-toolkits/skeleton-bridge/tree/master/sample)** (KendoUI, Materialize and this "skeleton bridge" get access to the **[plugin](https://github.com/aurelia-ui-toolkits/skeleton-bridge/tree/master/src)** in the `index.html` file that resides at the root level -- **NOT** in the `main.js`

Let me know if this needs additional explanations

@Karthick-NR please let me know why are using this repo>

**Comment by Thanood** [comment]
@adriatic Can you please tell me where this behaviour is defined in the code?
It must be new, since:
- [KendoUI](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/sample/src/main.js#L9)
- [Materialize](https://github.com/aurelia-ui-toolkits/aurelia-materialize-bridge/blob/master/sample/src/main.js#L5)
- [Syncfusion](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/blob/master/sample/src/main.js#L7)

are all using this line.

**Comment by adriatic** [comment]
@Thanood **Great catch** - Let's see how to recover from it. Here is a bit of history:
Initially, the catalog app and the bridge (plugin) were in different repositories, making the worklow:

```
Change the bridge
Build the bridge
Build and run the catalog app to verify the correctness of the bridge behavior
```

more complex than what it should be. So, @charlespockert added a nice trick with this code in catalog app's `index.html` file

```
    <script src="config.js"></script>
    <script>
      System.import('aurelia-bootstrapper');

      // Load the package.json from the root to get the package name
      System.import("root/package.json!").then(function(m) {
        var cfg = {
           packages: {}
        };

        // Create a new systemjs package with the package name
        cfg.packages[m.name] = {
            "main": "index",
            "dependencies": m.jspm.dependencies
        };

        System.config(cfg);

      });

    </script>
```

The "trick" is the idea to read the name of the repository as defined in the `package.json` file

![image](https://cloud.githubusercontent.com/assets/2712405/14607560/fc20c414-054f-11e6-9955-db8dba14fd86.png)

and build the configuration data dynamically - feature that resulted with the elimination of the need to declare this specific dependency in `main.js` file.

Then, we faced the issue of building two versions of the bridge (Kendo-core versus Kendo-PRO) - at which time (Jeroen, I believe) used `main.js` to solve this problem.

Finally the line `System.import("root/package.json!").then(function(m) {` became invalid without anyone noticing, until @karthickthangasamy today reported it as `Issue 1` **[above](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issue-149116536)**

Despite all this being a storm in a teapot, I am thankful to @Karthick-NR to unearth this.

**Comment by Thanood** [comment]
Thank you for this comprehensive explanation. :smile: 
I still don't understand _why_ it fails to load, though. I mean what exactly causes this since all the source files seem correct and `/src/root/package.json` should be accessible.

But I can find out myself, I think. I'm just curious (as always). No need to put any more time into this. :smile:

**Comment by adriatic** [comment]
@Thanood - let's continue this discussion at https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/issues/462

**Comment by JeroenVinke** [comment]
> ### Problem solved - read on:
> 
> **[This line](https://github.com/aurelia-ui-toolkits/skeleton-bridge/blob/master/sample/src/main.js#L7)** should NOT be there - the correct version of `main.js` should be

That conclusion is incorrect as that line should be there. Otherwise the catalog never loads the plugin. The catalog is using the plugin exactly the same as any other application would. The only difference is the way SystemJS loads it. Normal applications run `jspm install some-bridge-name` and SystemJS would load the plugin files from jspm_packages, while our catalog has a special mapping that happens in `index.html`.

A few other remarks:

> So, @charlespockert added a nice trick with this code in catalog app's index.html file
> @PWKad came up with that neat trick :)

I believe this line:

`System.import("root/package.json!").then(function(m) {`

is only used in order to retrieve the name of the plugin. That way, the name of the plugin is stored in a single place (package.json).  With it, the special mapping I talked about earlier is created with SystemJS.

**Comment by adriatic** [comment]
> That conclusion is incorrect as that line should be there. 

Two comments: 
1. My observation how this `index.html` based creation of config.js does not include any references to the plugin is really dumb as this config.js is created dynamically - not on the disk file, so I could not see that it does not do anything unless I use the debugger :-(
2. I still do not see how `System.import("root/package.json!").then(function(m) {` work! What is the value of `root` and where is it defined?

**Comment by JeroenVinke** [comment]
Because we are not installing (`jspm install aurelia-kendoui-bridge`) the bridge in the embedded sample app, we do not have [these lines](https://github.com/aurelia-ui-toolkits/demo-kendo/blob/master/config.js#L155-L163) in the config.js of the embedded sample app. So, the dynamic, at run time configuration equivalent is this:

![image](https://cloud.githubusercontent.com/assets/2189477/14610212/36b52eac-058d-11e6-9177-3b279babee1f.png)

This is necessary because when you call [this line](https://github.com/aurelia-ui-toolkits/demo-kendo/blob/master/src/main.js#L9), it will ask SystemJS to load some files from `aurelia-kendoui-bridge` and we have to register that name with SystemJS. 

So basically [these lines](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/sample/index.html#L26-L39) equals [these lines](https://github.com/aurelia-ui-toolkits/demo-kendo/blob/master/config.js#L155-L163).

Another issue has to do with the wwwroot of the gulp webserver:
![image](https://cloud.githubusercontent.com/assets/2189477/14609881/f433a87a-058b-11e6-99ae-ab43fa97db31.png)

The wwwroot is the `sample` folder and the webserver won't (just like other webservers) serve any files from outside that wwwroot.

[this line](https://github.com/aurelia-ui-toolkits/skeleton-bridge/blob/master/build/tasks/serve.js#L23) creates a mapping with the webserver:

`'/src/aurelia-kendoui-bridge` -> `devbuild/system/`

So that when you request `aurelia-kendoui-bridge/config-builder.js` the webserver will serve `config-builder.js` from outside its wwwroot from `devbuild/system/config-builder.js`

**Comment by adriatic** [comment]
Simply wonderful!

**Comment by bharathm03** [comment]
@adriatic We further worked on it to find the root cause of this build failure and the issue you mentioned in this [comment](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/issues/9#issuecomment-210180679) is related to latest browserSync version "2.12.3", when I set the browserSync version to "2.11.2", bridge is working fine.

Please refer the related issue link below : 
https://github.com/BrowserSync/browser-sync/issues/1063

And we are not using the skeleton bridge for development, We brought this issue here because we faced similar failure in syncfusion bridge also and found that it is related to skeleton-bridge itself.

**Comment by adriatic** [comment]
Thank you @bharathm03 - is seems that things are progressing in a good way. BrowserSync error is not what I expected to be the issue, so congrats on catching this. I will clone the Syncfusion bridge later today to give you overall feedback on all issues we discussed so far.

**Comment by adriatic** [comment]
@bharathm03 @JeroenVinke 

> We further worked on it to find the root cause of this build failure and the issue you mentioned in this [comment](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/issues/9#issuecomment-210180679) is related to latest browserSync version "2.12.3", when I set the browserSync version to "2.11.2", bridge is working fine.

I cannot verify that the **bridge is working fine**, so I  want to ask @JeroenVinke to step in and do us all a favor by doing this simple test
1. Get a clone of https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge
2. Build the catalog app and the bridge
3. run `gulp watch`

When I do the above steps, I get 

```
system.src.js:4597 GET http://localhost:3000/src/root/package.json 404 (Not Found)I @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597
system.js:4 Uncaught (in promise) Error: XHR error (404 Not Found) loading http://localhost:3000/src/root/package.json(…)t @ system.src.js:4597g @ system.src.js:4597(anonymous function) @ system.src.js:4597
aurelia-logging-console.js:38 DEBUG [aurelia] Loading plugin http://localhost:3000/jspm_packages/npm/aurelia-templating-binding@1.0.0-beta.1.1.2.
aurelia-logging-console.js:38 DEBUG [aurelia] Configured plugin http://localhost:3000/jspm_packages/npm/aurelia-templating-binding@1.0.0-beta.1.1.2.
aurelia-logging-console.js:38 DEBUG [aurelia] Loading plugin http://localhost:3000/jspm_packages/npm/aurelia-templating-resources@1.0.0-beta.1.1.3.
aurelia-logging-console.js:38 DEBUG [aurelia] Configured plugin http://localhost:3000/jspm_packages/npm/aurelia-templating-resources@1.0.0-beta.1.1.3.
aurelia-logging-console.js:38 DEBUG [aurelia] Loading plugin http://localhost:3000/jspm_packages/npm/aurelia-history-browser@1.0.0-beta.1.1.4.
aurelia-logging-console.js:38 DEBUG [aurelia] Configured plugin http://localhost:3000/jspm_packages/npm/aurelia-history-browser@1.0.0-beta.1.1.4.
aurelia-logging-console.js:38 DEBUG [aurelia] Loading plugin http://localhost:3000/jspm_packages/npm/aurelia-templating-router@1.0.0-beta.1.1.2.
aurelia-logging-console.js:38 DEBUG [aurelia] Configured plugin http://localhost:3000/jspm_packages/npm/aurelia-templating-router@1.0.0-beta.1.1.2.
aurelia-logging-console.js:38 DEBUG [aurelia] Loading plugin http://localhost:3000/jspm_packages/npm/aurelia-event-aggregator@1.0.0-beta.1.1.1.
aurelia-logging-console.js:38 DEBUG [aurelia] Configured plugin http://localhost:3000/jspm_packages/npm/aurelia-event-aggregator@1.0.0-beta.1.1.1.
aurelia-logging-console.js:38 DEBUG [aurelia] Loading plugin aurelia-syncfusion-bridge.
system.src.js:4597 GET http://localhost:3000/src/aurelia-syncfusion-bridge.js 404 (Not Found)I @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597(anonymous function) @ system.src.js:4597
system.js:4 Uncaught (in promise) Error: XHR error (404 Not Found) loading http://localhost:3000/src/aurelia-syncfusion-bridge.js(…)t @ system.src.js:4597g @ system.src.js:4597(anonymous function) @ system.src.js:4597
```

Error `system.src.js:4597 GET http://localhost:3000/src/root/package.json 404 (Not Found)` is caused by a bad path `http://localhost:3000/src/root/package.json` - as this should be 

```
http://localhost:3000/src/package.json
```

**Note 1**: I presented this same problem **[here](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issuecomment-211432934)** by writing:

2 I still do not see how `System.import("root/package.json!").then(function(m) {` work! What is the value of `root` and where is it defined?

**Note 2**: 
Error `system.src.js:4597 GET http://localhost:3000/src/aurelia-syncfusion-bridge.js 404 (Not Found)` is obviously the consequence of the first problem (failing to read the `package.json`) at which points nothing that follows makes any sense.

I am running node v5.1.1 and npm 3.5.0 - saying this as a possible difference in behavior that I experience trying to run Syncfusion catalog app (It **never** worked for me) and the experience declared bt @Thanood and @bharathm03 .

@JeroenVinke - can you get us straight, please?

**Comment by bharathm03** [comment]
@adriatic It seems like, you didn't changed the browerSync version to 2.11.2 before npm install. Now i committed the version correction in repo. Can you try again running the aurelia-syncfusion-bridge catelog?

**Comment by adriatic** [comment]
@bharathm03 this problem (that I described **[above](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issuecomment-212213643)**) takes place at the earliest moments in the application bootstrap, much, much sooner than BrowserSync gets activated. Specifically, I am talking about **[this code section](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/blob/master/sample/index.html#L22-L35)**

The same code section in Aurelia-KendoUI bridge executes correctly - so this ought to be a problem in the initial definition of various paths. I spend some time looking around this and found nothing. So, let's see what @JeroenVinke might find - he will likely show up in a couple of hours.

**Comment by bharathm03** [comment]
Ok @adriatic we will wait for @JeroenVinke . Also I like to share my findings, 
Cause of this issue is that [server routes](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issuecomment-212213643) defined in [this code section ](https://github.com/aurelia-ui-toolkits/skeleton-bridge/blob/master/build/tasks/serve.js#L16) is broken in browserSync. 

As per browserSync documentation [in this page](https://www.browsersync.io/docs/options/#option-server), it supposed to be relative to current working directory, which is breaking in recent version.

**Comment by JeroenVinke** [comment]
I just confirmed that [this commit](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/commit/67789e0e64142b1237954670385758abda303fd2) fixed the issue mentioned [here](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issuecomment-212213643). I saw the error with  `^2.9.3` and it worked with `2.11.2`.

**Comment by adriatic** [comment]
#### Several comments
1. @bharathm03 **[original explanation above](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issuecomment-211929136)** is **correct!**
2. My own **[above written littany](https://github.com/aurelia-ui-toolkits/skeleton-bridge/issues/26#issuecomment-212213643)**  is based on the fact that the definition of the **[BrowserSync version](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/package.json#L67)** KendoUI bridge is that same `"browser-sync": "^2.9.3"`, that crashes Syncfusion bridge.
3. I understand now that I do not understand what is going on here - something is still not right. Let's keep this under observation.
4. My apologies to @bharathm03 for not taking his explanation directly.

**In summary**: Syncfusion bridge works and looks terrific and I now know that we have a clear path forward.

**Comment by adriatic** [comment]
One more important comment: I mentioned earlier that the current implementation of the wrappers for **[`button`](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/tree/master/src/button)** and **[`menu`](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge/tree/master/src/menu)** are incomplete: they both have only the code for the constructor - all other callbacks (`bind`, `attached`, and `detached` as well as `recreate` method are missing.

Please see the **[kendo wrapper](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/src/button/button.js)** for an example.

**Comment by bharathm03** [comment]
Thanks @adriatic , @JeroenVinke and @Thanood

**Comment by adriatic** [comment]
As this issue should really be taken place in **[Syncfusion repository](https://github.com/aurelia-ui-toolkits/aurelia-syncfusion-bridge)** I will close it here and referenced it from there.
